File: mediapipe/tasks/web/vision/face_stylizer/face_stylizer.ts
/**
 * Copyright 2022 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {CalculatorGraphConfig} from '../../../../framework/calculator_pb';
import {CalculatorOptions} from '../../../../framework/calculator_options_pb';
import {BaseOptions as BaseOptionsProto} from '../../../../tasks/cc/core/proto/base_options_pb';
import {FaceStylizerGraphOptions as FaceStylizerGraphOptionsProto} from '../../../../tasks/cc/vision/face_stylizer/proto/face_stylizer_graph_options_pb';
import {WasmFileset} from '../../../../tasks/web/core/wasm_fileset';
import {MPImage} from '../../../../tasks/web/vision/core/image';
import {ImageProcessingOptions} from '../../../../tasks/web/vision/core/image_processing_options';
import {
  VisionGraphRunner,
  VisionTaskRunner,
} from '../../../../tasks/web/vision/core/vision_task_runner';
import {
  ImageSource,
  WasmModule,
} from '../../../../web/graph_runner/graph_runner';
// Placeholder for internal dependency on trusted resource url

import {FaceStylizerOptions} from './face_stylizer_options';

export * from './face_stylizer_options';
export {type ImageSource}; // Used in the public API

const IMAGE_STREAM = 'image_in';
const NORM_RECT_STREAM = 'norm_rect';
const STYLIZED_IMAGE_STREAM = 'stylized_image';
const FACE_STYLIZER_GRAPH =
  'mediapipe.tasks.vision.face_stylizer.FaceStylizerGraph';

// The OSS JS API does not support the builder pattern.
// tslint:disable:jspb-use-builder-pattern

/**
 * A callback that receives an `MPImage` object from the face stylizer, or
 * `null` if no face was detected. The lifetime of the underlying data is
 * limited to the duration of the callback. If asynchronous processing is
 * needed, all data needs to be copied before the callback returns (via
 * `image.clone()`).
 */
export type FaceStylizerCallback = (image: MPImage | null) => void;

/** Performs face stylization on images. */
export class FaceStylizer extends VisionTaskRunner {
  private userCallback?: FaceStylizerCallback;
  private result?: MPImage | null;
  private readonly options: FaceStylizerGraphOptionsProto;

  /**
   * Initializes the Wasm runtime and creates a new Face Stylizer from the
   * provided options.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param faceStylizerOptions The options for the Face Stylizer. Note
   *     that either a path to the model asset or a model buffer needs to be
   *     provided (via `baseOptions`).
   */
  static createFromOptions(
    wasmFileset: WasmFileset,
    faceStylizerOptions: FaceStylizerOptions,
  ): Promise<FaceStylizer> {
    return VisionTaskRunner.createVisionInstance(
      FaceStylizer,
      wasmFileset,
      faceStylizerOptions,
    );
  }

  /**
   * Initializes the Wasm runtime and creates a new Face Stylizer based on
   * the provided model asset buffer.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param modelAssetBuffer An array or a stream containing a binary
   *    representation of the model.
   */
  static createFromModelBuffer(
    wasmFileset: WasmFileset,
    modelAssetBuffer: Uint8Array | ReadableStreamDefaultReader,
  ): Promise<FaceStylizer> {
    return VisionTaskRunner.createVisionInstance(FaceStylizer, wasmFileset, {
      baseOptions: {modelAssetBuffer},
    });
  }

  /**
   * Initializes the Wasm runtime and creates a new Face Stylizer based on
   * the path to the model asset.
   * @export
   * @param wasmFileset A configuration object that provides the location of
   *     the Wasm binary and its loader.
   * @param modelAssetPath The path to the model asset.
   */
  static createFromModelPath(
    wasmFileset: WasmFileset,
    modelAssetPath: string,
  ): Promise<FaceStylizer> {
    return VisionTaskRunner.createVisionInstance(FaceStylizer, wasmFileset, {
      baseOptions: {modelAssetPath},
    });
  }

  /** @hideconstructor */
  constructor(
    wasmModule: WasmModule,
    glCanvas?: HTMLCanvasElement | OffscreenCanvas | null,
  ) {
    super(
      new VisionGraphRunner(wasmModule, glCanvas),
      IMAGE_STREAM,
      NORM_RECT_STREAM,
      /* roiAllowed= */ true,
    );
    this.options = new FaceStylizerGraphOptionsProto();
    this.options.setBaseOptions(new BaseOptionsProto());
  }

  protected override get baseOptions(): BaseOptionsProto {
    return this.options.getBaseOptions()!;
  }

  protected override set baseOptions(proto: BaseOptionsProto) {
    this.options.setBaseOptions(proto);
  }

  /**
   * Sets new options for the Face Stylizer.
   *
   * Calling `setOptions()` with a subset of options only affects those
   * options. You can reset an option back to its default value by
   * explicitly setting it to `undefined`.
   *
   * @export
   * @param options The options for the Face Stylizer.
   */
  override setOptions(options: FaceStylizerOptions): Promise<void> {
    return super.applyOptions(options);
  }

  /**
   * Performs face stylization on the provided single image and invokes the
   * callback with result. The method returns synchronously once the callback
   * returns. Only use this method when the FaceStylizer is created with the
   * image running mode.
   *
   * @param image An image to process.
   * @param callback The callback that is invoked with the stylized image or
   *    `null` if no face was detected. The lifetime of the returned data is
   *     only guaranteed for the duration of the callback.
   */
  stylize(image: ImageSource, callback: FaceStylizerCallback): void;
  /**
   * Performs face stylization on the provided single image and invokes the
   * callback with result. The method returns synchronously once the callback
   * returns. Only use this method when the FaceStylizer is created with the
   * image running mode.
   *
   * The 'imageProcessingOptions' parameter can be used to specify one or all
   * of:
   *  - the rotation to apply to the image before performing stylization, by
   *    setting its 'rotationDegrees' property.
   *  - the region-of-interest on which to perform stylization, by setting its
   *   'regionOfInterest' property. If not specified, the full image is used.
   *  If both are specified, the crop around the region-of-interest is extracted
   *  first, then the specified rotation is applied to the crop.
   *
   * @param image An image to process.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input image before running inference.
   * @param callback The callback that is invoked with the stylized image or
   *    `null` if no face was detected. The lifetime of the returned data is
   *    only guaranteed for the duration of the callback.
   */
  stylize(
    image: ImageSource,
    imageProcessingOptions: ImageProcessingOptions,
    callback: FaceStylizerCallback,
  ): void;
  /**
   * Performs face stylization on the provided single image and returns the
   * result. This method creates a copy of the resulting image and should not be
   * used in high-throughput applications. Only use this method when the
   * FaceStylizer is created with the image running mode.
   *
   * @param image An image to process.
   * @return A stylized face or `null` if no face was detected. The result is
   *     copied to avoid lifetime issues.
   */
  stylize(image: ImageSource): MPImage | null;
  /**
   * Performs face stylization on the provided single image and returns the
   * result. This method creates a copy of the resulting image and should not be
   * used in high-throughput applications. Only use this method when the
   * FaceStylizer is created with the image running mode.
   *
   * The 'imageProcessingOptions' parameter can be used to specify one or all
   * of:
   *  - the rotation to apply to the image before performing stylization, by
   *    setting its 'rotationDegrees' property.
   *  - the region-of-interest on which to perform stylization, by setting its
   *   'regionOfInterest' property. If not specified, the full image is used.
   *  If both are specified, the crop around the region-of-interest is extracted
   *  first, then the specified rotation is applied to the crop.
   *
   * @param image An image to process.
   * @param imageProcessingOptions the `ImageProcessingOptions` specifying how
   *    to process the input image before running inference.
   * @return A stylized face or `null` if no face was detected. The result is
   *     copied to avoid lifetime issues.
   */
  stylize(
    image: ImageSource,
    imageProcessingOptions: ImageProcessingOptions,
  ): MPImage | null;
  /** @export */
  stylize(
    image: ImageSource,
    imageProcessingOptionsOrCallback?:
      | ImageProcessingOptions
      | FaceStylizerCallback,
    callback?: FaceStylizerCallback,
  ): MPImage | null | void {
    const imageProcessingOptions =
      typeof imageProcessingOptionsOrCallback !== 'function'
        ? imageProcessingOptionsOrCallback
        : {};

    this.userCallback =
      typeof imageProcessingOptionsOrCallback === 'function'
        ? imageProcessingOptionsOrCallback
        : callback;
    this.processImageData(image, imageProcessingOptions ?? {});

    if (!this.userCallback) {
      return this.result;
    }
  }

  /** Updates the MediaPipe graph configuration. */
  protected override refreshGraph(): void {
    const graphConfig = new CalculatorGraphConfig();
    graphConfig.addInputStream(IMAGE_STREAM);
    graphConfig.addInputStream(NORM_RECT_STREAM);
    graphConfig.addOutputStream(STYLIZED_IMAGE_STREAM);

    const calculatorOptions = new CalculatorOptions();
    calculatorOptions.setExtension(
      FaceStylizerGraphOptionsProto.ext,
      this.options,
    );

    const segmenterNode = new CalculatorGraphConfig.Node();
    segmenterNode.setCalculator(FACE_STYLIZER_GRAPH);
    segmenterNode.addInputStream('IMAGE:' + IMAGE_STREAM);
    segmenterNode.addInputStream('NORM_RECT:' + NORM_RECT_STREAM);
    segmenterNode.addOutputStream('STYLIZED_IMAGE:' + STYLIZED_IMAGE_STREAM);
    segmenterNode.setOptions(calculatorOptions);

    graphConfig.addNode(segmenterNode);

    this.graphRunner.attachImageListener(
      STYLIZED_IMAGE_STREAM,
      (wasmImage, timestamp) => {
        const mpImage = this.convertToMPImage(
          wasmImage,
          /* shouldCopyData= */ !this.userCallback,
        );
        this.result = mpImage;
        if (this.userCallback) {
          this.userCallback(mpImage);
        }
        this.setLatestOutputTimestamp(timestamp);
      },
    );
    this.graphRunner.attachEmptyPacketListener(
      STYLIZED_IMAGE_STREAM,
      (timestamp) => {
        this.result = null;
        if (this.userCallback) {
          this.userCallback(null);
        }
        this.setLatestOutputTimestamp(timestamp);
      },
    );

    const binaryGraph = graphConfig.serializeBinary();
    this.setGraph(new Uint8Array(binaryGraph), /* isBinary= */ true);
  }
}

File: mediapipe/tasks/python/vision/face_stylizer.py
# Copyright 2023 The MediaPipe Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""MediaPipe face stylizer task."""

import dataclasses
from typing import Optional

from mediapipe.python import packet_creator
from mediapipe.python import packet_getter
from mediapipe.python._framework_bindings import image as image_module
from mediapipe.tasks.cc.vision.face_stylizer.proto import face_stylizer_graph_options_pb2
from mediapipe.tasks.python.core import base_options as base_options_module
from mediapipe.tasks.python.core import task_info as task_info_module
from mediapipe.tasks.python.core.optional_dependencies import doc_controls
from mediapipe.tasks.python.vision.core import base_vision_task_api
from mediapipe.tasks.python.vision.core import image_processing_options as image_processing_options_module
from mediapipe.tasks.python.vision.core import vision_task_running_mode as running_mode_module

_BaseOptions = base_options_module.BaseOptions
_FaceStylizerGraphOptionsProto = (
    face_stylizer_graph_options_pb2.FaceStylizerGraphOptions
)
_ImageProcessingOptions = image_processing_options_module.ImageProcessingOptions
_TaskInfo = task_info_module.TaskInfo

_STYLIZED_IMAGE_NAME = 'stylized_image'
_STYLIZED_IMAGE_TAG = 'STYLIZED_IMAGE'
_NORM_RECT_STREAM_NAME = 'norm_rect_in'
_NORM_RECT_TAG = 'NORM_RECT'
_IMAGE_IN_STREAM_NAME = 'image_in'
_IMAGE_OUT_STREAM_NAME = 'image_out'
_IMAGE_TAG = 'IMAGE'
_TASK_GRAPH_NAME = 'mediapipe.tasks.vision.face_stylizer.FaceStylizerGraph'
_MICRO_SECONDS_PER_MILLISECOND = 1000


@dataclasses.dataclass
class FaceStylizerOptions:
  """Options for the face stylizer task.

  Attributes:
    base_options: Base options for the face stylizer task.
  """

  base_options: _BaseOptions

  @doc_controls.do_not_generate_docs
  def to_pb2(self) -> _FaceStylizerGraphOptionsProto:
    """Generates an FaceStylizerOptions protobuf object."""
    base_options_proto = self.base_options.to_pb2()
    return _FaceStylizerGraphOptionsProto(base_options=base_options_proto)


class FaceStylizer(base_vision_task_api.BaseVisionTaskApi):
  """Class that performs face stylization on images."""

  @classmethod
  def create_from_model_path(cls, model_path: str) -> 'FaceStylizer':
    """Creates an `FaceStylizer` object from a TensorFlow Lite model and the default `FaceStylizerOptions`.

    Note that the created `FaceStylizer` instance is in image mode, for
    stylizing one face on a single image input.

    Args:
      model_path: Path to the model.

    Returns:
      `FaceStylizer` object that's created from the model file and the default
      `FaceStylizerOptions`.

    Raises:
      ValueError: If failed to create `FaceStylizer` object from the provided
        file such as invalid file path.
      RuntimeError: If other types of error occurred.
    """
    base_options = _BaseOptions(model_asset_path=model_path)
    options = FaceStylizerOptions(base_options=base_options)
    return cls.create_from_options(options)

  @classmethod
  def create_from_options(cls, options: FaceStylizerOptions) -> 'FaceStylizer':
    """Creates the `FaceStylizer` object from face stylizer options.

    Args:
      options: Options for the face stylizer task.

    Returns:
      `FaceStylizer` object that's created from `options`.

    Raises:
      ValueError: If failed to create `FaceStylizer` object from
        `FaceStylizerOptions` such as missing the model.
      RuntimeError: If other types of error occurred.
    """

    task_info = _TaskInfo(
        task_graph=_TASK_GRAPH_NAME,
        input_streams=[
            ':'.join([_IMAGE_TAG, _IMAGE_IN_STREAM_NAME]),
            ':'.join([_NORM_RECT_TAG, _NORM_RECT_STREAM_NAME]),
        ],
        output_streams=[
            ':'.join([_STYLIZED_IMAGE_TAG, _STYLIZED_IMAGE_NAME]),
            ':'.join([_IMAGE_TAG, _IMAGE_OUT_STREAM_NAME]),
        ],
        task_options=options,
    )
    return cls(
        task_info.generate_graph_config(),
        running_mode=running_mode_module.VisionTaskRunningMode.IMAGE,
    )

  def stylize(
      self,
      image: image_module.Image,
      image_processing_options: Optional[_ImageProcessingOptions] = None,
  ) -> image_module.Image:
    """Performs face stylization on the provided MediaPipe Image.

    Only use this method when the FaceStylizer is created with the image
    running mode.

    Args:
      image: MediaPipe Image.
      image_processing_options: Options for image processing.

    Returns:
      The stylized image of the most visible face. The stylized output image
      size is the same as the model output size. None if no face is detected
      on the input image.

    Raises:
      ValueError: If any of the input arguments is invalid.
      RuntimeError: If face stylization failed to run.
    """
    normalized_rect = self.convert_to_normalized_rect(
        image_processing_options, image
    )
    output_packets = self._process_image_data({
        _IMAGE_IN_STREAM_NAME: packet_creator.create_image(image),
        _NORM_RECT_STREAM_NAME: packet_creator.create_proto(
            normalized_rect.to_pb2()
        ),
    })
    if output_packets[_STYLIZED_IMAGE_NAME].is_empty():
      return None
    return packet_getter.get_image(output_packets[_STYLIZED_IMAGE_NAME])
